datastream-server/src/main/java/com/linkedin/datastream/server/EventProducer.java

```diff
@@ -342,6 +347,13 @@ private void performSlaRelatedLogging(DatastreamRecordMetadata metadata, long ev
   private void reportMetrics(DatastreamRecordMetadata metadata, long eventsSourceTimestamp, long eventsSendTimestamp) {
     // If per-topic metrics are enabled, use topic as key for metrics; else, use datastream name as the key
     String datastreamName = getDatastreamName();
+
+    Set<String> throughputViolatingTopics = _throughputViolatingTopicsProvider.apply(_datastreamTask);
+    if (throughputViolatingTopics.contains(metadata.getTopic())) {
+      _logger.warn(
```

Comment:
You will print this warning for every record. That is very expensive and unnecessary.